Descent3/mission_download.cpp

```diff
@@ -338,181 +334,161 @@ int msn_DownloadWithStatus(char *url, char *filename) {
-                                  ret = msn_ExtractZipFile(qualfile,filename);
-                          }else
-                          {
-                                  ret = 1;
-                          }
-                  }
-
-                  if(getmsnfile->IsFileError())
-                  {
-                          //File transfer Error!
-                          DoMessageBox(TXT_ERROR,TXT_FMTCANTDNLD,MSGBOX_OK);
-                          //Delete the file that didn't finish!
-                          ddio_DeleteFile(qualfile);
-                          mprintf(0,"Couldn't download the file! Error: %d\n",getmsnfile->GetErrorCode());
-                          exit_menu = 1;
-                          ret = 0;
-                  }
+  cancel_hot.Create(&menu_wnd, UID_CANCEL, KEY_ESC, &cancel_off, &cancel_on, cancel_x,
+                    MSN_DWNLD_STATUS_H - OKCANCEL_YOFFSET, 0, 0, UIF_FIT | UIF_CENTER);
+  menu_wnd.Open();
 
-                  last_refresh = timer_GetTime();
-                  received_bytes = getmsnfile->GetBytesIn();
-                  total_bytes = getmsnfile->GetTotalBytes();
+  last_refresh = timer_GetTime() - MSN_REFRESH_INTERVAL;
 
-                  time_elapsed = timer_GetTime()-starttime;
+  std::string download_uri = "/" + StringJoin(std::vector<std::string>(url_parts.begin() + 3, url_parts.end()), "/");
 
-                  if(total_bytes)
-                  {
-                          time_remain = ((float)(total_bytes-received_bytes))/((float)(received_bytes/time_elapsed));
-                  }
-                  if(time_elapsed&&received_bytes)
-                  {
-                          xfer_rate = ((float)(received_bytes/time_elapsed));
-                  }
-                  texts[1].Destroy();
-                  texts[2].Destroy();
-                  texts[3].Destroy();
-                  texts[4].Destroy();
-                  texts[5].Destroy();
-                  texts[6].Destroy();
+  while (!exit_menu) {
+    int res;
 
-                  sprintf(fmturl,DOWNLOAD_STATUS_URL_TEXT,url);
-                  msn_ClipURLToWidth(MSN_DWNLD_STATUS_W-(MSN_COL_1+MSN_BORDER_W),fmturl);
-                  sprintf(fmtrcvd,DOWNLOAD_STATUS_RCVD_TEXT,received_bytes);
-                  sprintf(fmttotal,DOWNLOAD_STATUS_TOTAL_TEXT,total_bytes);
-                  sprintf(fmtelaps,DOWNLOAD_STATUS_ELAPS_TEXT,msn_SecondsToString(time_elapsed));
-                  sprintf(fmttimer,DOWNLOAD_STATUS_TIME_R_TEXT,msn_SecondsToString(time_remain));
-                  msn_ClipURLToWidth(MSN_DWNLD_STATUS_W-(MSN_COL_2+MSN_BORDER_W),fmttimer);
-                  sprintf(fmtrate,DOWNLOAD_STATUS_XFERRATE_TEXT,xfer_rate);
+    std::fstream in;
+    in.open(qualfile, std::ios::binary | std::ios::trunc | std::ios::out);
+    int status = http_client->Get(download_uri, in, [&received_bytes, &total_bytes](uint64_t len, uint64_t total) {
+      received_bytes = len;
+      total_bytes = total;
```

Comment:
the callback in the [lib readme example](https://github.com/yhirose/cpp-httplib?tab=readme-ov-file#with-progress-callback) uses a side-effect (print) to update on the request progress. Here this callback has none, I'm confused how this is supposed to work, because there is no asynchronicity

Reply:
`received_bytes` and `total_bytes` is outside variables that can updated in lambda function. Later these two are used for calculation ETA and speed.